Add device posture test driver and web platform tests

JuhaVainio opened the pull request.

JuhaVainio committed:
Add device posture tests

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,89 @@
+  });
+
+  const postures = ['folded', 'continuous', 'folded'];
+  for (let posture of postures) {
```
```suggestion
  for (const posture of postures) {
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/d2c3a71b9a1408a4c661ef228308296479937607

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,89 @@
+  const postures = ['folded', 'continuous', 'folded'];
+  for (let posture of postures) {
+    await test_driver.set_device_posture(posture);
+    await promise;
```
You can't `await` for the same promise multiple times. After it resolves once, you're just getting the same result it gave you the first time.

JuhaVainio replied:
Because of request of using EventWatcher , this part of code changed and no promise is anymore used.

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,89 @@
+
+promise_test(async (t) => {
+  t.add_cleanup(cleanup);
+  const promise = new Promise(resolve => {
```
Use an EventWatcher instead.

JuhaVainio replied:
This is not the code but in this case what I should use in EventWatcher?

const watcher = new EventWatcher(t, navigator.devicePosture, ['change']);
await test_driver.set_device_posture('folded');
watcher.wait_for('change').then(() => {
  console.log(navigator.devicePosture.type);
  watcher.stop();
});

Witrh above code I get "assert_true: Not expecting event, but got change event expected true got false".

JuhaVainio replied:
I think it might be more like this

  const watcher = new EventWatcher(t, navigator.devicePosture, ['change']);
  /* **await removed** */ test_driver.set_device_posture('folded');
  **await** watcher.wait_for('change').then(() => {
    console.log('JV666 1700 ' + navigator.devicePosture.type);
  });

JuhaVainio replied:
EventWatcher used in https://github.com/JuhaVainio/WPT/pull/4/commits/b1b8d4766e1c998e8e4e41a78a97ce5bf6f4e2c3

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+async function cleanup() {
+  await test_driver.set_device_posture('continuous');
```
This will always reset the posture to the same value even if it was something else by default. You need to cache the original value and restore it instead.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/e0de4c478f6e9e490ea2fd0e03655c99e69d4dd6

JuhaVainio committed:
In cleanup use original device posture value and not hard coded

JuhaVainio committed:
Define variable as constant when appropriate

JuhaVainio committed:
Replace promise with EventWatcher

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+//<script src="../resources/orientation-event-helpers.js"></script>
```
This is not the right syntax for HTML comments.

JuhaVainio replied:
Line removed in https://github.com/JuhaVainio/WPT/pull/4/commits/0ff7d73903eda5ad4a795743a15a623ab5f9ca4b

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+<script>
+'use strict';
+
+const originalPosture = navigator.devicePosture.type;
```
Caching this once outside all tests is not the way to go. This also requires every test author to remember to do this in their files, which is not very helpful or scalable.

I think something like what [`window_state_context.js`](https://github.com/web-platform-tests/wpt/blob/master/page-visibility/resources/window_state_context.js) does fit the model better here: keep some state around, add a thin wrapper for test_driver.set_device_posture() and automatically clean up if necessary.

JuhaVainio replied:
Some clarification needed

So have file e.g. device-posture/resources/device_posture_context.js

Which content would be 
```
function device_posture_context(t) {
  const originalPosture = navigator.devicePosture.type;
  t.add_cleanup(async () => {
    if (navigator.devicePosture.type !== originalPosture) {
      await test_driver.set_device_posture(originalPosture);
    }
  });
}
```

Each test would have this kind of first lines
```
promise_test(async (t) => {
  device_posture_context(t);

```

The "_thin wrapper for test_driver.set_device_posture()_" doesn't compute... or I do not see what is the need for wrapping/modify test_driver.set_device_posture() then?

rakuco replied:
The main difference is that window_state_context() returns a group of functions and yours doesn't. What I had in mind was a wrapper just like window_state_context's minimize() is a wrapper around `test_driver.minimize_window()` with extra bookkeeping.

Thinking of it again, the model is slightly different for device posture, as the window state code basically toggles between 2 states and the same function (e.g. minimize, or restore) isn't normally called more than once without calling the other.

What you could do instead is something more similar to the old implementation of the `sensor_test()` function before the WebDriver conversion work:

```js
function device_posture_test(func, name, properties) {
  promise_test(async t => {
    // save current posture
    // add cleanup function
    return func(t);
  }, name, properties);
}
```

JuhaVainio replied:
Cleanup fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/23376d3c2c8300e415829911eaf03261e6d20ac3

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+  const watcher = new EventWatcher(t, navigator.devicePosture, ['change']);
+  const postures = ['folded', 'continuous', 'folded'];
+  for (const posture of postures) {
+    test_driver.set_device_posture(posture);
```
This function returns a Promise that you need to handle (i.e. you are neither using then() or `await`ing on it).

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/1b0487df6e9b6b543ff6c28a2d28571ceb8b9398

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+  const postures = ['folded', 'continuous', 'folded'];
+  for (const posture of postures) {
+    test_driver.set_device_posture(posture);
+    await watcher.wait_for('change').then(() => {
```
You are mixing two different programming models: you either `await` for a Promise or you handle it via then().

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/578dc80038ba8a75b52bf886ed40e7b63907371b

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+
+
+// REFERENCE https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/inspector-protocol/device-posture/device-posture-change-event.js
+promise_test(async (t) => {
```
I don't see what this test is doing that the first one isn't.

JuhaVainio replied:
So this test is better as it is referenced from https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/inspector-protocol/device-posture/device-posture-change-event.js?

rakuco replied:
I think the one you wrote is more complete than this port of the DevTools test that only exercises one kind of posture.

JuhaVainio replied:
This specific test was removed in https://github.com/JuhaVainio/WPT/pull/4/commits/a5ae20689448dff656a4b5582e333cd1db126519, and work continues on other commits

rakuco commented on device-posture/device-posture.https.html:
```diff
@@ -0,0 +1,86 @@
+      { once: true }
+    );
+  });
+  test_driver.set_device_posture('folded');
```
You're not handling the promise here either.

JuhaVainio replied:
What does "not handling the promise" in here means? 
Isn't "await promise" in line 83 waiting the event to happen. And also if "resolve(foldedMQL.matches);" line is executed ==> the promise is handled?

rakuco replied:
"The promise" isn't the same as "`promise`". You are handling `promise` in line 83 by `await`ing on it, but I'm referring to _the promise_ returned by `test_driver.set_device_posture()` in this line. In the current version of the code, nothing guarantees that this promise has resolved by the time the test finishes and that it has run successfully.

JuhaVainio replied:
OK , the promise of test_driver.set_device_posture call. 

Learning the asynchronous nature of this language/Chromium env is ongoing. I was thinking that if you wait test_driver.set_device_posture to complete the event "assert_true(await promise)" is waiting would happen before we are in that line (so it would "miss" the event). Also "assert_true(await promise)" would guarantee that test cannot finish without device-posture to be  folded.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/3831e1ba726cd5fad98c5de482ade8d24222fe12. The first test will be fixed in another commit

rakuco commented on resources/testdriver.js:
```diff
@@ -1023,7 +1023,15 @@
         get_virtual_sensor_information: function(sensor_type, context=null) {
             return window.test_driver_internal.get_virtual_sensor_information(sensor_type, context);
-        }
+        },
```
Notice how the last item didn't have a comma but your additions do.

JuhaVainio replied:
Comma removed in https://github.com/JuhaVainio/WPT/pull/4/commits/0ff7d73903eda5ad4a795743a15a623ab5f9ca4b

rakuco commented on tools/wptrunner/wptrunner/executors/protocol.py:
```diff
@@ -802,3 +802,13 @@ def remove_virtual_sensor(self, sensor_type):
+
+    @abstractmethod
+    def set_device_posture(self, posture):
+        pass
```
Add a newline at the end of the file.

JuhaVainio replied:
Newline added in https://github.com/JuhaVainio/WPT/pull/4/commits/0ff7d73903eda5ad4a795743a15a623ab5f9ca4b

rakuco reviewed (changes requested):
There are programming issues in the converted tests that must be fixed. I've added inline comments pointing them out.

You should also use EventWatcher everywhere but the one test that wants to exercise addEventListener behavior.

JuhaVainio committed:
Fix coding convention errors

Missing newline from tools/wptrunner/wptrunner/executors/protocol.py.

Remove comma from resources/testdriver.js.

Unused commented code in device-posture/device-posture.https.html.

JuhaVainio committed:
Do not require each test to set up cleanup function

Removed the requirement from all tests to set up cleanup function.

All tests are now wrapped around device_posture_test call. This enables
that adding of cleanup function for each test is done in one location.

JuhaVainio committed:
Combine two event style tests

Two tests where similar, so other was delete.

JuhaVainio committed:
Wait promise from test_driver.set_device_posture call

JuhaVainio committed:
Wait promise from test_driver.set_device_posture also in first test

JuhaVainio committed:
Fix the wrong styte of waiting EventWatcher's promise

EventWatcher promise was waited with mixed style. Code was fixed to use
only one style of waiting. After change EventWatcher eventTypes
parameter value needed to change also to able to run test correctly.

JuhaVainio committed:
Remove "window" word from set device posture URI

JuhaVainio committed:
Add viewport web test

JuhaVainio committed:
Rename visual-viewport to display-feature

Also added rest of the parameters which display-feature needs.

JuhaVainio committed:
Remove display feature parts

JuhaVainio committed:
Split test to own files. Code clean up.

JuhaVainio commented on resources/testdriver.js:
```diff
@@ -1203,6 +1210,10 @@
+
+        async set_device_posture(posture, context=null) {
+            throw new Error("set_device_posture() is not implemented by testdriver-vendor.js");
         }
```
Add clear_device_posture

JuhaVainio replied:
Added in 2ebcd5d750

JuhaVainio commented on device-posture/device-posture-change-event.https.html:
```diff
@@ -0,0 +1,19 @@
+// REFERENCE https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/inspector-protocol/device-posture/device-posture-change-event.js
+promise_test(async (t) => {
+  const watcher = new EventWatcher(t, navigator.devicePosture, ['onchange']);
+  const postures = ['continuous', 'folded', 'continuous', 'folded'];
```
Fails if code is `const postures = ['folded', 'continuous', 'folded'];`
Fails on first loop by saying `navigator.devicePosture.type` is `continuous` and should be `folded`.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/291385c15ed5cf39ad5a37182bcc1994ee799744

JuhaVainio committed:
Add clear device posture and test for it

JuhaVainio committed:
Document set_device_posture and clear_device_posture in testdriver.js

JuhaVainio committed:
Add set_device_posture and clear_device_posture to docs/writing-tests/testdriver.md

JuhaVainio committed:
Add set_device_posture and clear_device_posture to tools/wptrunner/wptrunner/executors/executormarionette.py

JuhaVainio committed:
Add "Protocol" word to WebDriverDevicePosturePart in executorwebdriver.py

Every other class name has "Protocol" word

rakuco commented on device-posture/device-posture-change-event.https.html:
```diff
@@ -0,0 +1,19 @@
+
+// REFERENCE https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/inspector-protocol/device-posture/device-posture-change-event.js
+promise_test(async (t) => {
+  const watcher = new EventWatcher(t, navigator.devicePosture, ['onchange']);
```
"onchange" is the wrong event name. Look at other EventWatcher users do ("foo" and "onfoo" are different).

JuhaVainio replied:
> Look at other EventWatcher users do ("foo" and "onfoo" are different).

Unfortunately this tip is not very useful. The variation between different EventWatcher's is huge. They wait load, message, securitypolicyviolation, activate, reading, error, ...

I think I should look https://w3c.github.io/device-posture/ . I was confused as it talks lot about "onchange" (and that test seems to works (partially) with "onchange" word), but now I found this chapter https://w3c.github.io/device-posture/#the-onchange-attribute-handle-posture-changes which says the event type is "change"

After onchange => change I get into problems with this line
`await test_driver.set_device_posture(posture);`
If I have "await" in there I get "Not expecting event, but got change event expected true got false" error. 
Also I have to change the "await watcher;" line
With onchange => change and with below kind of look code seems to work

```
promise_test(async (t) => {
  const watcher = new EventWatcher(t, navigator.devicePosture, ['change']);
  const postures = ['folded', 'continuous', 'folded'];
  for (const posture of postures) {
    test_driver.set_device_posture(posture);
    await watcher.wait_for('change');
    assert_equals(posture, navigator.devicePosture.type);
  }
}, 'Device posture test: device-posture-change-event.js.');
```

rakuco replied:
> Unfortunately this tip is not very useful. The variation between different EventWatcher's is huge. They wait load, message, securitypolicyviolation, activate, reading, error, ...
> 
> I think I should look https://w3c.github.io/device-posture/ . I was confused as it talks lot about "onchange" (and that test seems to works (partially) with "onchange" word), but now I found this chapter https://w3c.github.io/device-posture/#the-onchange-attribute-handle-posture-changes which says the event type is "change"

I did not write a more detailed explanation because this shouldn't have been a surprise to you by now. You've written code that does `object.onfoo = ...` as well as `object.addEventListener('foo')`, and you've dealt with EventWatcher in Generic Sensors and Device Orientation in the past months. The Generic Sensors spec has [a section listing event handlers](https://w3c.github.io/sensors/#event-handlers), and the entire Device Orientation API is based around events, so a `git grep EventWatcher` in the corresponding directories in WPT should have given you more information to support this work.


> After onchange => change I get into problems with this line `await test_driver.set_device_posture(posture);` If I have "await" in there I get "Not expecting event, but got change event expected true got false" error. Also I have to change the "await watcher;" line With onchange => change and with below kind of look code seems to work
> 
> ```
> promise_test(async (t) => {
>   const watcher = new EventWatcher(t, navigator.devicePosture, ['change']);
>   const postures = ['folded', 'continuous', 'folded'];
>   for (const posture of postures) {
>     test_driver.set_device_posture(posture);
>     await watcher.wait_for('change');
>     assert_equals(posture, navigator.devicePosture.type);
>   }
> }, 'Device posture test: device-posture-change-event.js.');
> ```

This is a variation on a familiar theme (e.g. `test_driver.update_virtual_sensor()` and `watcher.wait_for('reading')`), so similar advice applies.

```js
// Option 1.
await Promise.all([
  watcher.wait_for('change'),
  test_driver.set_device_posture(posture)
]);

// Option 2, just a more verbose version of option 1.
const watchPromise = watcher.wait_for('change');
await test_driver.set_device_posture(posture);
await watchPromise;
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/291385c15ed5cf39ad5a37182bcc1994ee799744

rakuco commented on device-posture/device-posture-change-event.https.html:
```diff
@@ -0,0 +1,19 @@
+<script>
+'use strict';
+
+// REFERENCE https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/inspector-protocol/device-posture/device-posture-change-event.js
```
Remove these comments.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/cb93f7e142b6128add4d546d6e11142f42b7b48c

rakuco commented on device-posture/device-posture-clear.https.html:
```diff
@@ -0,0 +1,19 @@
+  const posture = originalPosture ? 'folded' : 'continuous';
+
+  await test_driver.set_device_posture(posture);
+  assert_equals(navigator.devicePosture.type, posture);
```
Wait for the "change" event before checking `type`.

JuhaVainio replied:
Fixed in f09d788860

rakuco commented on device-posture/device-posture-default.https.html:
```diff
@@ -0,0 +1,15 @@
+  assert_equals(navigator.devicePosture.type, 'continuous');
+  await test_driver.set_device_posture('folded');
+  assert_equals(navigator.devicePosture.type, 'folded')
+}, 'Device posture test: device-posture-default.js.');
```
No need to prefix the descriptions with "Device posture test".

JuhaVainio replied:
Reference link removed and test descriptions fixed in https://github.com/JuhaVainio/WPT/pull/4/commits/cb93f7e142b6128add4d546d6e11142f42b7b48c

rakuco reviewed (commented):
Not sure if you were waiting for my review, but I've added some comments about things that need to be fixed in addition to anything else that was already on your list.

JuhaVainio committed:
Fix wrong event type

Event type was changed from 'onchange' to 'change'. This led change in
waiting for event. Which solved the problem that 'continuous' posture
had to be the first posture.

JuhaVainio committed:
Remove reference links and copy the test description from reference test

JuhaVainio committed:
Add waiting of change when setting or clearing device posture in clear test

JuhaVainio committed:
Remove device-posture-default.https.html test

Removed because other tests already had same code and read posture
through navigator.devicePosture.type.

JuhaVainio committed:
Fix code format issues

rakuco reviewed (commented):
One thing that's just occurred to me while working on the content_shell CL upstream: all these tests that call set_device_posture() need a

```js
t.add_cleanup(async () => {
  await test_driver.clear_device_posture();
});
```

call to make sure nothing leaks between tests.
